### PR TITLE
Support zone

### PR DESCRIPTION
**Pull request by darionyaphet**

### CODE REVIEW

File: src/meta/KVBasedClusterIdMan.h [hunk 1]
```diff
@@ -106,16 +106,16 @@ class ClusterIdMan {
...
-                           });
+                          if (code != kvstore::ResultCode::SUCCEEDED) {
+                              LOG(ERROR) << "Put failed, error "
+                                         << static_cast<int32_t>(code);
+                              ret = false;
+                          } else {
+                              LOG(INFO) << "Put key " << key
+                                        << ", val " << clusterId;
+                          }
+                              baton.post();
```
Comment (critical27) [inline src/meta/KVBasedClusterIdMan.h:117]: alignment

File: src/meta/processors/zoneMan/AddGroupProcessor.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    if (!nebula::ok(groupIdRet)) {
+        LOG(ERROR) << "Create Group failed";
+        handleErrorCode(nebula::error(groupIdRet));
+        onFinished();
+        return;
+    }
+
+    std::vector<kvstore::KV> data;
+    auto groupId = nebula::value(groupIdRet);
+    data.emplace_back(MetaServiceUtils::indexGroupKey(groupName),
```
Comment (critical27) [inline src/meta/processors/zoneMan/AddGroupProcessor.cpp:55]: group id and zone id is only used to check whether name conflict, right?

File: src/meta/processors/zoneMan/AddGroupProcessor.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                      std::string(reinterpret_cast<const char*>(&groupId), sizeof(GroupID)));
+    data.emplace_back(MetaServiceUtils::groupKey(groupName),
+                      MetaServiceUtils::groupVal(zoneNames));
+
+    LOG(INFO) << "Create Group: " << groupName;
+    doSyncPutAndUpdate(std::move(data));
+}
+
+bool AddGroupProcessor::checkGroupRedundancy(std::vector<std::string> zones) {
```
Comment (critical27) [inline src/meta/processors/zoneMan/AddGroupProcessor.cpp:64]: `const std::vector<std::string>&`? BTW, do we need to check this, I think group is more like a alias of several zones. It is ok to have multiple same alias.

File: src/meta/processors/zoneMan/DropZoneProcessor.cpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        handleErrorCode(cpp2::ErrorCode::E_NOT_FOUND);
+        onFinished();
+        return;
+    }
+
+    std::vector<std::string> keys;
+    keys.emplace_back(MetaServiceUtils::indexZoneKey(zoneName));
+    keys.emplace_back(MetaServiceUtils::zoneKey(zoneName));
+    LOG(INFO) << "Drop Zone " << zoneName;
+    doSyncMultiRemoveAndUpdate(std::move(keys));
```
Comment (critical27) [inline src/meta/processors/zoneMan/DropZoneProcessor.cpp:27]: Need to check all hosts in zone is empty?

File: src/meta/processors/zoneMan/AddGroupProcessor.cpp (hunk 2 above)
Comment (darionyaphet) [reply src/meta/processors/zoneMan/AddGroupProcessor.cpp:55]: Currently they are use to check duplicate but later maybe act as the internal representation

File: src/meta/processors/zoneMan/AddGroupProcessor.cpp (hunk 3 above)
Comment (darionyaphet) [reply src/meta/processors/zoneMan/AddGroupProcessor.cpp:64]: I'm afraid the two group create with the same zones although the order is different. Such as create group_0 z-0,z-1,z-2  and create group_1 z-2,z-1,z-0

File: src/meta/processors/zoneMan/DropZoneProcessor.cpp (hunk 4 above)
Comment (darionyaphet) [reply src/meta/processors/zoneMan/DropZoneProcessor.cpp:27]: maybe we should check whether the zone belong to any group and the group is related to a space?

File: src/meta/MetaServiceUtils.cpp [hunk 5]
```diff
@@ -475,16 +477,36 @@ std::string MetaServiceUtils::indexEdgeKey(GraphSpaceID spaceId,
...
+       .append(reinterpret_cast<const char*>(&type), sizeof(type))
        .append(reinterpret_cast<const char*>(&spaceID), sizeof(GraphSpaceID))
        .append(indexName);
     return key;
 }
 
+std::string MetaServiceUtils::indexGroupKey(const std::string& name) {
+    EntryType type = EntryType::GROUP;
+    std::string key;
+    key.reserve(128);
```
[the remark this answers is not in the document]
Comment (darionyaphet) [reply src/meta/MetaServiceUtils.cpp:493]: yep but this is unified with another index key

**Review by critical27 — APPROVED** [review]
Generally LGTM

**Review by liuyu85cn — APPROVED** [review]
LGTM

**Pull request merged**